src/main/java/org/opensearch/security/util/TransportNodesAsyncAction.java

```diff
@@ -0,0 +1,305 @@
+            if (this.concreteNodes.length == 0) {
+                // nothing to notify
+                threadPool.generic().execute(() -> listener.onResponse(newResponse(request, responses)));
+                return;
+            }
+            TransportRequestOptions.Builder builder = TransportRequestOptions.builder();
+            if (request.timeout() != null) {
+                builder.withTimeout(request.timeout());
+            }
+            for (int i = 0; i < this.concreteNodes.length; i++) {
+                final int idx = i;
+                final DiscoveryNode node = this.concreteNodes[i];
+                final String nodeId = node.getId();
+                try {
+                    TransportRequest nodeRequest = newNodeRequest(request);
+                    if (task != null) {
+                        nodeRequest.setParentTask(clusterService.localNode().getId(), task.getId());
+                    }
+                    transportService.sendRequest(
+                        node,
+                        getTransportNodeAction(node),
+                        nodeRequest,
+                        builder.build(),
+                        new TransportResponseHandler<NodeResponse>() {
+                            @Override
+                            public NodeResponse read(StreamInput in) throws IOException {
+                                return newNodeResponse(in);
+                            }
+
+                            @Override
+                            public void handleResponse(NodeResponse response) {
+                                onOperation(idx, response);
+                            }
+
+                            @Override
+                            public void handleException(TransportException exp) {
+                                onFailure(idx, node.getId(), exp);
+                            }
+
+                            @Override
+                            public String executor() {
+                                return ThreadPool.Names.SAME;
+                            }
+                        }
+                    );
+                } catch (Exception e) {
+                    onFailure(idx, nodeId, e);
+                }
+            }
+        }
+
+        private void onOperation(int idx, NodeResponse nodeResponse) {
+            responses.set(idx, nodeResponse);
+            if (counter.incrementAndGet() == responses.length()) {
+                finishHim();
+            }
+        }
+
+        private void onFailure(int idx, String nodeId, Throwable t) {
+            if (logger.isDebugEnabled() && !(t instanceof NodeShouldNotConnectException)) {
+                logger.debug(new ParameterizedMessage("failed to execute on node [{}]", nodeId), t);
+            }
+            responses.set(idx, new FailedNodeException(nodeId, "Failed node [" + nodeId + "]", t));
+            if (counter.incrementAndGet() == responses.length()) {
+                finishHim();
```

Comment:
nit: rename it to finishExecutor?

Reply:
yeah, I also do not like that wording, I took it over from the core `TransportNodesAction`. If we move this over to core, we might have to keep the wording if we want to be compatible, though.